### PR TITLE
Add __slots__ to expressions

### PR DESCRIPTION
**Pull request by ehennenfent**
This PR modifies SMTLib's inheritance structure to avoid using multiple inheritance. This allows us to use `__slots__` to improve memory usage and performance with SMTLib expressions.

## Old inheritance diagram:
Old Diagram

## New Inheritance diagram:
Constant, Variable, and Operation have been inlined in their subclasses

The `Constant`, `Variable`, and `Operation` classes from the old diagram have been inlined into their respective subclasses. I've added a test to make sure that the subclasses still expose the same methods. I've also added tuples containing all the subclasses (eg `Constant = (BitVecConstant, BoolConstant)`) so that `isinstance(x, Constant)` will still work. 

Running this on `multiple_styles`, I saw a 3.5% increase in speed and a 4.5% decrease in overall memory usage. But of course, `multiple-styles` is a very lightweight example.

### CODE REVIEW

File: manticore/core/smtlib/expression.py [hunk 1]
```diff
@@ -215,26 +162,62 @@ def __bool__(self):
...
-        super().__init__(value, *args, **kwargs)
+        self._value = value
+        super().__init__(*args, **kwargs)
 
     def __bool__(self):
         return self.value
 
+    @property
+    def value(self):
+        return self._value
```
Comment (disconnect3d) [inline manticore/core/smtlib/expression.py]: Ideally, the property should be there only in a debug mode :<

File: manticore/core/smtlib/expression.py [hunk 2]
```diff
@@ -215,26 +162,62 @@ def __bool__(self):
...
+    __slots__ = ["_operands"]
 
-class BoolOperation(Operation, Bool):
     def __init__(self, *operands, **kwargs):
-        super().__init__(*operands, **kwargs)
+        self._operands = operands
+
+        # If taint was not forced by a keyword argument, calculate default
+        if "taint" not in kwargs:
+            kwargs["taint"] = reduce(lambda x, y: x.union(y.taint), operands, frozenset())
```
Comment (disconnect3d) [inline manticore/core/smtlib/expression.py]: Could be...
```python
kwargs.setdefault('taint', reduce(lambda x, y: x.union(y.taint), operands, frozenset()))
```

File: manticore/core/smtlib/expression.py (hunk 1 above)
Comment (disconnect3d) [reply manticore/core/smtlib/expression.py]: Random thought: maybe we should use the private version in internal code? I *believe* (not confirmed) those attributes might be somehow *hot* and we don't want to use a call internally to get their value if possible

File: manticore/core/smtlib/expression.py [hunk 3]
```diff
@@ -465,18 +450,38 @@ def Bool(self):
...
+        return self._name
+
+    def __copy__(self, memo):
+        raise ExpressionException("Copying of Variables is not allowed.")
+
+    def __deepcopy__(self, memo):
+        raise ExpressionException("Copying of Variables is not allowed.")
+
+    def __repr__(self):
+        return "<{:s}({:s}) at {:x}>".format(type(self).__name__, self.name, id(self))
```
Comment (disconnect3d) [inline manticore/core/smtlib/expression.py]: I wonder if there is much difference between `type(self)` and `self.__class__`

File: manticore/core/smtlib/expression.py [hunk 4]
```diff
@@ -1204,3 +1266,8 @@ def __init__(
         assert true_value.size == size
         assert false_value.size == size
         super().__init__(size, condition, true_value, false_value, *args, **kwargs)
+
+
+Constant = (BitVecConstant, BoolConstant)
+Variable = (BitVecVariable, BoolVariable, ArrayVariable)
+Operation = (BitVecOperation, BoolOperation, ArrayOperation, ArraySelect)
```
Comment (disconnect3d) [inline manticore/core/smtlib/expression.py]: Maybe `Constants`, `Variables`, `Operations`?

File: manticore/core/smtlib/constraints.py [hunk 5]
```diff
@@ -158,7 +158,7 @@ def to_string(self, related_to=None, replace_constants=False):
                 if (
                     isinstance(expression, BoolEqual)
                     and isinstance(expression.operands[0], Variable)
-                    and isinstance(expression.operands[1], (Variable, Constant))
+                    and isinstance(expression.operands[1], (*Variable, *Constant))
```
Comment (disconnect3d) [inline manticore/core/smtlib/constraints.py]: or `Variables + Constants`, heh

File: manticore/core/smtlib/expression.py (hunk 1 above)
Comment (ehennenfent) [reply manticore/core/smtlib/expression.py]: Previous maintainers agree with you :
https://github.com/trailofbits/manticore/blob/abb191a3a562f30aba42ec4b4b4b4604c283130f/manticore/core/smtlib/visitors.py#L779
We should strive to present the same API to users, but we could probably replace a lot of internal property uses with direct access.

File: manticore/core/smtlib/expression.py (hunk 4 above)
Comment (ehennenfent) [reply manticore/core/smtlib/expression.py]: 97647b33dbad014cbe914afdc610046794a585e7

File: manticore/core/smtlib/visitors.py [hunk 6]
```diff
@@ -861,17 +930,26 @@ def visit_BitVecConstant(self, expression):
     def visit_BoolConstant(self, expression):
         return expression.value and "true" or "false"
 
-    def visit_Variable(self, expression):
+    def _visit_variable(self, expression):
         return expression.name
 
+    def visit_ArrayVariable(self, expression):
```
Comment (disconnect3d) [inline manticore/core/smtlib/visitors.py]: Could also be:
```python
In [10]: class X:
    ...:     def _priv(self, x):
    ...:         return x
    ...:     public = _priv
    ...:

In [11]: X().public(3)
Out[11]: 3

In [12]: X().public.__name__
Out[12]: '_priv'
```

but names :<

**Pull request merged**